Cleaning C types

Typical panics after analyzing by `csmith` in type casting:
```
volatile union U3 ** * --> volatile union U3 **volatile *

panic: invalid identity: 'U3 **SliceToU3 *Slice' [recovered]
        panic: bad type: 'noarch.U3 **SliceToU3 *Slice'

goroutine 1 [running]:
github.com/elliotchance/c2go/util.typeToExpr.func1(0xc42122a020, 0x1c)
        /home/lepricon/go/src/github.com/elliotchance/c2go/util/goast.go:41 +0x115

```
```
union U5 ** * --> union U5 ***

panic: invalid identity: 'U5 **SliceToU5 *Slice' [recovered]
        panic: bad type: 'noarch.U5 **SliceToU5 *Slice'

goroutine 1 [running]:
github.com/elliotchance/c2go/util.typeToExpr.func1(0xc4206b2a60, 0x1c)
        /home/lepricon/go/src/github.com/elliotchance/c2go/util/goast.go:41 +0x115
panic(0x5a1480, 0xc4218a53d0)
```
```
const struct S2 * * --> const struct S2 **

panic: invalid identity: 'S2 *SliceToS2Slice' [recovered]
        panic: bad type: 'noarch.S2 *SliceToS2Slice'

goroutine 1 [running]:
github.com/elliotchance/c2go/util.typeToExpr.func1(0xc420a0bf40, 0x19)
        /home/lepricon/go/src/github.com/elliotchance/c2go/util/goast.go:41 +0x115
```
```
struct S1 * * --> struct S1 **

panic: invalid identity: 'S1 *SliceToS1Slice' [recovered]
        panic: bad type: 'noarch.S1 *SliceToS1Slice'

goroutine 1 [running]:
github.com/elliotchance/c2go/util.typeToExpr.func1(0xc420798300, 0x19)
        /home/lepricon/go/src/github.com/elliotchance/c2go/util/goast.go:41 +0x115
```

--- a/ast/function_decl_test.go
+++ b/ast/function_decl_test.go
@@ -166,6 +166,22 @@ func TestFunctionDecl(t *testing.T) {
 			IsInline:     true,
 			ChildNodes:   []Node{},
 		},
+		`0x21c3da0 <line:8201:1, line:8786:1> line:8201:25 used insertvertex 'enum insertvertexresult (struct mesh *, struct behavior *, vertex, struct otri *, struct osub *, int, int)'`: &FunctionDecl{
+			Addr:         0x21c3da0,
+			Pos:          NewPositionFromString("line:8201:1, line:8786:1"),
+			Prev:         "",
+			Position2:    "line:8201:25",
+			Name:         "insertvertex",
+			Type:         "enum insertvertexresult (struct mesh *, struct behavior *, vertex, struct otri *, struct osub *, int, int)",
+			Type2:        "",
+			IsExtern:     false,
+			IsImplicit:   false,
+			IsUsed:       true,
+			IsReferenced: false,
+			IsStatic:     false,
+			IsInline:     false,
+			ChildNodes:   []Node{},
+		},
 	}
 
 	runNodeTests(t, nodes)

--- a/types/cast.go
+++ b/types/cast.go
@@ -61,6 +61,9 @@ func GetArrayTypeAndSize(s string) (string, int) {
 //    FILE where those function probably exist (or should exist) in the noarch
 //    package.
 func CastExpr(p *program.Program, expr goast.Expr, cFromType, cToType string) (goast.Expr, error) {
+	cFromType = CleanCType(cFromType)
+	cToType = CleanCType(cToType)
+
 	fromType := cFromType
 	toType := cToType
 

--- a/types/resolve.go
+++ b/types/resolve.go
@@ -322,10 +322,10 @@ func CleanCType(s string) (out string) {
 	out = strings.Replace(out, "(*)", "", -1)
 
 	// Remove any whitespace or attributes that are not relevant to Go.
-	out = strings.Replace(out, "const", " ", -1)
-	out = strings.Replace(out, "volatile", " ", -1)
-	out = strings.Replace(out, "__restrict", " ", -1)
-	out = strings.Replace(out, "restrict", " ", -1)
+	out = strings.Replace(out, "const", "", -1)
+	out = strings.Replace(out, "volatile", "", -1)
+	out = strings.Replace(out, "__restrict", "", -1)
+	out = strings.Replace(out, "restrict", "", -1)
 	out = strings.Replace(out, "\t", "", -1)
 	out = strings.Replace(out, "\n", "", -1)
 	out = strings.Replace(out, "\r", "", -1)

--- a/types/resolve.go
+++ b/types/resolve.go
@@ -111,13 +111,7 @@ var simpleResolveTypes = map[string]string{
 //    transpiler to step over type errors and put something as a placeholder
 //    until a more suitable solution is found for those cases.
 func ResolveType(p *program.Program, s string) (string, error) {
-	// Remove any whitespace or attributes that are not relevant to Go.
-	s = strings.Replace(s, "const ", "", -1)
-	s = strings.Replace(s, "volatile ", "", -1)
-	s = strings.Replace(s, "*__restrict", "*", -1)
-	s = strings.Replace(s, "*restrict", "*", -1)
-	s = strings.Replace(s, "*const", "*", -1)
-	s = strings.Trim(s, " \t\n\r")
+	s = CleanCType(s)
 
 	// FIXME: This is a hack to avoid casting in some situations.
 	if s == "" {
@@ -319,4 +313,35 @@ func ParseFunction(s string) (f []string, r []string, err error) {
 	}
 
 	return
+}
+
+// CleanCType - remove from C type not Go type
+func CleanCType(s string) (out string) {
+	out = s
+	out = strings.Replace(out, "()", "", -1)
+	out = strings.Replace(out, "(*)", "", -1)
+
+	// Remove any whitespace or attributes that are not relevant to Go.
+	out = strings.Replace(out, "const", " ", -1)
+	out = strings.Replace(out, "volatile", " ", -1)
+	out = strings.Replace(out, "__restrict", " ", -1)
+	out = strings.Replace(out, "restrict", " ", -1)
+	out = strings.Replace(out, "\t", "", -1)
+	out = strings.Replace(out, "\n", "", -1)
+	out = strings.Replace(out, "\r", "", -1)
+
+	// remove space from pointer symbols
+	out = strings.Replace(out, "* *", "**", -1)
+
+	// remove addition spaces
+	out = strings.Replace(out, "  ", " ", -1)
+
+	// remove spaces around
+	out = strings.TrimSpace(out)
+
+	if out != s {
+		return CleanCType(out)
+	}
+
+	return out
 }

